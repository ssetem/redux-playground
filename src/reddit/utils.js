export function createAction(type, ...argNames) {
    let action = function (...args) {
        let action = { type }
        argNames.forEach((arg, index) => {
            action[argNames[index]] = args[index]
        })
        return action
    }
    action.type = type
    return action
}

export function createReducer(initialState, handlers) {
    return function reducer(state = initialState, action) {
        if (handlers.hasOwnProperty(action.type)) {
            return handlers[action.type](state, action)
        } else {
            return state
        }
    }
}