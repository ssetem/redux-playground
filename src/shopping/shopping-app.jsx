import * as React from "react"
import { Provider, connect } from "react-redux"
import { configureStore, outlets } from "./shopping"

let store = configureStore()



let Cart = ({ items, total, removeProduct }) => {
    return (
        <div>
            <table>
                <thead>
                    <th>Product</th>
                    <th>Quantity</th>
                    <th>Price</th>
                    <th></th>
                </thead>
                <tbody>

                    {items.map((item) => {
                        return (
                            <tr>
                                <td>{item.product.name}</td>
                                <td>{item.quantity}</td>
                                <td>£{item.totalPrice}</td>
                                <td><span onClick={()=> removeProduct(item.product.id)}>x</span></td>
                            </tr>
                        )
                    })}
                    <tr>
                        <td></td>
                        <td>total:</td>
                        <td>£{total}</td>
                    </tr>
                </tbody>
            </table>
          
        </div>
    )
}

Cart = connect(...outlets.cart)(Cart)


let Products = ({ products, addProduct }) => {
    return (
        <ul>
            {products.map((product) => {
                return <li
                    onClick={()=> addProduct(product.id)} 
                    key={product.id}>{product.name} £{product.price}</li>
            })}
        </ul>
    )
}

Products = connect(...outlets.products)(Products)

export default () => {
    return (
        <Provider store={store}>
            <div>
                <Products />
                <Cart />
            </div>
        </Provider>
    )
}
