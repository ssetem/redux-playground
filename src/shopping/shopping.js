import {
    combineReducers, createStore,
    applyMiddleware, compose
} from "redux"
import { 
    createAction, createReducer, singleReduce, decrement,
    actionProp, union, indexByKey, increment, remove
} from "../utils/Utils"
import _ from "lodash"
import thunk from "redux-thunk"
import { createSelector } from "reselect"

const property = require("lodash/property")

export const types = {
    ADD_PRODUCT: "ADD_PRODUCT",
    REMOVE_PRODUCT: "REMOVE_PRODUCT"
}
export const actions = {
    addProduct: createAction(types.ADD_PRODUCT, "productId"),
    removeProduct: createAction(types.REMOVE_PRODUCT, "productId")
}

const initialProducts = {
    byId: {
        "1": { id: "1", name: "iPhone X", price: 1000 },
        "2": { id: "2", name: "Galaxy Note 8", price: 800 },
        "3": { id: "3", name: "Macbook Pro 13", price: 1400 }
    },
    ids: ["1", "2", "3"]
}

const products = createReducer(initialProducts, {})

const cart = indexByKey("productId", [types.ADD_PRODUCT, types.REMOVE_PRODUCT],
    createReducer(0, {
        [types.ADD_PRODUCT]: increment,
        [types.REMOVE_PRODUCT]: decrement
    })
)
export const reducers = combineReducers({
    products,
    cart
})


const getCartItems = createSelector(
    property("cart"),property("products.byId"),
    (cart, products) => {
        let total = 0        
        let items = _.compact(Object.keys(cart).map((id) => {
            let product = products[id]
            let quantity = cart[id]
            if(quantity < 1){
                return
            }

            let totalPrice = product.price * quantity
            total += totalPrice
            return {
                quantity, product, totalPrice
            }
        }))
        return {items, total}
    }
)
const getProductList = createSelector(
    property("products.ids"), property("products.byId"),
    (ids, byIds)=> ids.map((id)=> byIds[id])
    
)
export const selectors = { getCartItems, getProductList }

export const outlets = {
    cart: [
        (state) => {
            return getCartItems(state)                
        },
        { 
            addProduct: actions.addProduct,
            removeProduct: actions.removeProduct 
        }
    ],

    products:[
        (state)=> ({
          products:getProductList(state)
        }), 
        { addProduct: actions.addProduct }
    ]
}

export const configureStore = (initialState = {}) => {
    return createStore(
        reducers,
        initialState,
        compose(
            applyMiddleware(thunk),
            window.__REDUX_DEVTOOLS_EXTENSION__ ? window.__REDUX_DEVTOOLS_EXTENSION__() : _.identity
        )
    )
}

