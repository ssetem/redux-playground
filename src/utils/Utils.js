import shallowequal from "shallowequal"
import difflet from "difflet"
import { bindActionCreators } from "redux"
let _ = require("lodash")

export function createAction(type, ...argNames) {
    let action = function (...args) {
        let action = { type }
        argNames.forEach((arg, index) => {
            action[argNames[index]] = args[index]
        })
        return action
    }
    action.type = type
    return action
}

export function createReducer(initialState, handlers) {
    return function reducer(state = initialState, action) {
        if (handlers.hasOwnProperty(action.type)) {
            return handlers[action.type](state, action)
        } else {
            return state
        }
    }
}

export const actionProp = (key) => (state, action) => action[key]
export const increment = (state, action) => state + 1
export const decrement = (state, action) => state - 1
export const union = (fn) => (state, action) => _.union(state, [fn(state, action)])
export const add = (fn) => (state, action) => [...state, fn(state, action)]
export const remove = (fn) => (state, action) => _.without(state, [fn(state, action)])
export const eq = (fn) => (state, action) => state === fn(state, action)
export const nullIf = (fn) => (state, action) => fn(state, action) ? null : state

export const indexByKey = (key, actions, subreducer) => {
    return (state = {}, action) => {
        if (actions.indexOf(action.type) > -1) {
            return {
                ...state,
                [action[key]]: subreducer(state[action[key]], action)
            }
        }
        return state
    }
}

export const singleReduce = (initialState, type, fn) => {
    return (state = initialState, action) => {
        if (action.type === type) {
            return fn(state, action)
        } else {
            return state
        }
    }
}

export const connectOutletToStore = (outlet, store, options={ownProps:{}})=> {
    const [mapStateToProps, mapDispatchToProps={}] = outlet
    let previousProps = {}
    let statefulOutlet = {        
        ownProps:options.ownProps,
        props:{},
        actions:{},    
        propsChanged:false    
    }
    const loadState = () => {
        let props = mapStateToProps(store.getState(), statefulOutlet.ownProps)
        previousProps = statefulOutlet.props
        statefulOutlet.props = props
        statefulOutlet.propsChanged = !shallowequal(previousProps, statefulOutlet.props)
        statefulOutlet.actions = bindActionCreators(mapDispatchToProps, store.dispatch)
    }

    statefulOutlet.setOwnProps = (ownProps)=> {
        statefulOutlet.ownProps = {
            ...statefulOutlet.ownProps, 
            ...ownProps
        }
        loadState()
    }

    statefulOutlet.getPreviousProps = ()=> {
        return previousProps
    }

    store.subscribe(loadState)
    loadState()
    return statefulOutlet
}

export const logOutletChanges = (outlet)=> {
    if(outlet.propsChanged) {
        console.log(difflet({ indent: 2, comment: true }).compare(
            outlet.getPreviousProps(), outlet.props
        ))   
    } else {
        console.log("No changes detected")
    }
    
}