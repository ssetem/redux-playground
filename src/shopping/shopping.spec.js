import { configureStore, actions, selectors, outlets } from "./shopping"
import { bindOutlet, connectOutletToStore, logOutletChanges } from "../utils/Utils"

describe("shopping", function () {

    fit("testing stateful outlets", ()=> {
        const store = configureStore()
        let cartOutlet = connectOutletToStore(outlets.cart, store)
        let productOutlet = connectOutletToStore(outlets.products, store)
        expect(cartOutlet.props).toMatchSnapshot()
        cartOutlet.actions.addProduct("1")        
        expect(cartOutlet.props).toMatchSnapshot()
        cartOutlet.actions.addProduct("1")        
        expect(cartOutlet.props).toMatchSnapshot()
        cartOutlet.actions.addProduct("2")        
        expect(cartOutlet.props).toMatchSnapshot()
        cartOutlet.actions.removeProduct("1")
        console.log(cartOutlet.getPreviousProps())  
        console.log(cartOutlet)        
        // cartOutlet.actions.removeProduct("1")
        expect(cartOutlet.props).toMatchSnapshot()
        cartOutlet.actions.removeProduct("1")
        expect(cartOutlet.props).toMatchSnapshot()
        cartOutlet.actions.removeProduct("2")
        console.log(cartOutlet.props)
        
    })

    it("test own props", ()=> {
        const store = configureStore()
        let productOutletByOwnprops = [
            (state, ownProps)=> {
                return {
                    product:state.products.byId[ownProps.productId]
                }
            }
        ]

        let outlet = bindOutletAndListen(
            productOutletByOwnprops, 
            store, 
            {ownProps:{productId:"1"}}
        )
        outlet.setOwnProps({productId:"2"})
        outlet.logChanges()
        console.log(outlet)
        
    })
})