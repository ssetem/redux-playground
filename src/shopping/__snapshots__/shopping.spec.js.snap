// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shopping testing stateful outlets 1`] = `
Object {
  "items": Array [],
  "total": 0,
}
`;

exports[`shopping testing stateful outlets 2`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "id": "1",
        "name": "iPhone X",
        "price": 1000,
      },
      "quantity": 1,
      "totalPrice": 1000,
    },
  ],
  "total": 1000,
}
`;

exports[`shopping testing stateful outlets 3`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "id": "1",
        "name": "iPhone X",
        "price": 1000,
      },
      "quantity": 2,
      "totalPrice": 2000,
    },
  ],
  "total": 2000,
}
`;

exports[`shopping testing stateful outlets 4`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "id": "1",
        "name": "iPhone X",
        "price": 1000,
      },
      "quantity": 2,
      "totalPrice": 2000,
    },
    Object {
      "product": Object {
        "id": "2",
        "name": "Galaxy Note 8",
        "price": 800,
      },
      "quantity": 1,
      "totalPrice": 800,
    },
  ],
  "total": 2800,
}
`;

exports[`shopping testing stateful outlets 5`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "id": "1",
        "name": "iPhone X",
        "price": 1000,
      },
      "quantity": 1,
      "totalPrice": 1000,
    },
    Object {
      "product": Object {
        "id": "2",
        "name": "Galaxy Note 8",
        "price": 800,
      },
      "quantity": 1,
      "totalPrice": 800,
    },
  ],
  "total": 1800,
}
`;

exports[`shopping testing stateful outlets 6`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "id": "2",
        "name": "Galaxy Note 8",
        "price": 800,
      },
      "quantity": 1,
      "totalPrice": 800,
    },
  ],
  "total": 800,
}
`;
